src/main/java/io/usethesource/vallang/util/WeakReferenceCache.java

```diff
@@ -0,0 +1,481 @@
+ */ 
+package io.usethesource.vallang.util;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.concurrent.locks.StampedLock;
+import java.util.function.Function;
+
+/**
+ * <p>
+ *     A cache that stores the either the key or the value (or both) as a weak reference. If either one of the references is cleared (in case of a weak reference), the entry is dropped from the cache.
+ *     The cache is thread safe, meaning the {@link #get} method can be called without any locking requirements.
+ *     With the following guarantees: 
+ * <p>
+ * <ul>
+ *    <li> As long as there are strong references to both the key and the value, all calls to {@link #get} will return the value reference for something that {@link Object#equals} key (not just reference equality, as Caffeine and WeakHashMap do)
+ *    <li> There will only be one entry for a given key, there is no point in time where you can get two different values for the same key (as the entry wasn't cleared because there was no strong reference to the key or the value between two calls)
+ *    <li> If the key is in the cache, retrieving the value will not block the code.
+ *    <li> If the key is *not* in the cache, the function to generate a value will be called once per concurrent {@link #get} call, and only one of those results will be the reference stored in the cache, and returned to all these concurrent calls.
+ *    <li> If the key is *not* in the cache, the {@link #get} rarely blocks, only in case of a resize (which happens at 80% fill rate, and grows exponentially)
+ * </ul>
+ * 
+ * <p>
+ *     Warning: only use this class if you want to clear the entry after either the value or the key has been cleared, if you do not want this behavior, use a regular {@link Map}.
+ *     Also check the {@link WeakHashMap} that keeps a strong reference to the Value, however it uses reference equality on the key.
+ * </p>
+ * @author Davy Landman
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class WeakReferenceCache<K,V> {
```

Comment:
Is it worth to implement this as a set rather than a map? to save memory I mean? since we use it only for the Flyweight pattern.

Reply:
I started out that way, but realized this class might be a bit more useful if we have it as a key-value container.  The Node class does avoid allocating two weak-references if the key is the same reference as the value.

I'm especially thinking that the memo feature of rascal could be using this class. Except that for IValue the equality and hashing are a bit to complex with annotations.